feat: escort policy (#2)
open-autonomy/escort: README.md

Reviewer: Suggest add section for:

## Driver Requirements

For the system to function as intended, the following driving behaviours must be adhered to:

**All parties shall:**
 - Adhere to the traffic rules specified by the map

**The 'Escorter' must:**
- Adhere to the speed limit specified at the time of creation
- Drive in such a manner as to ensure the 'Escortee/s' can follow within the 'Length' of the escort

**The 'Escortee' must:**
 - Always follow the path of the Escorter at a safe driving distance
 - Follow the Escorter, as closely as possible, within the bounds of a safe distance
 - When on roads, the Escorter shall follow within the boundaries of the lane
 - When in open areas, the Escorter shall follow within the boundaries of the 'Width' of the escort
 - If the Escorter stops, the Escortee shall pull up directly behind, within the bounds of a safe distance.

> [!WARNING]  
> Failure to adhere to the specified driving behaviours could result in loss of protection provided by the system.

Reviewer: Suggest to add section for:

## System Limitations
- Oversized vehicles are not supported

Oversized vehicles are defined as vehicles that meet any of the following conditions:
- Block both lanes of traffic
- Cannot remain within a single lane for the entire journey
- For any reason, cannot adhere to the road rules specified by the map

Reviewer 2: Suggestion https://github.com/open-autonomy/escort/pull/5

# Escort 
The Escort feature allows non-instrumented equipment to safely travel in the autonomous operating zone by being escorted by a staffed instrumented vehicle. The non-instrumented equipment is protected by always being inside a zone behind the instrumented vehicle, communicated to all autonomous vehicles.
 

> This repository defines the **messages and communication protocol between the Fleet Management System (FMS) and the Autonomous Haulage System (AHS)** for activating and deactivating escorts.

Reviewer 3: And providing updates on the escort position

 
## Terminology
| Term | Description |  
| --- | --- |  
| AHS | Autonomous Haulage System |  
| AV | Autonomous Vehicle |  
| FMS | Fleet Management System |  
| AOZ | Autonomous Operating Zone. The area where the autonomus mining vehicles are allowed to operate. |

> | SIV | Staffed Instrumented Vehicle. Manually driven equipment equipped with a remote stop. Could be a car, an excavator, or any other type of moving equipment. |

Reviewer: Suggest removing reference to remote stop. 

Possibly need a small section on the SIV that contains at least the following information:
- Typical use case expects the 'Escorter' SIV to use a light vehicle

Minimum requirements for the instrumentation are that it has the ability to:
- Determine its position with a precision measurement
- Network connection to the FMS

Optional:
- Connection to secondary network layer (e.g. v2x)

Reviewer 2: Suggestion https://github.com/open-autonomy/escort/pull/5

| Escorter | The leading SIV during the escort activity. |
| Escortee | A non-instrumented vehicle being escorted. |
| Escorting convoy | The escorter together with its escortee(s) |
| Escort Protection Zone | The zone in which the escorting convoy is required to be within at all times. Is defined by a configurable distance that extends behind the escorter, to include the escortee. |

> | Escort Avoidance Zone | Zone internally defined on each AV, that the AV will not enter. It is at least as big as the protection zone, but is likely bigger due to latency of the escorter's position updates. |

Reviewer: I think we need a section on the main page to specify this.

- Each AV manages its own avoidance zone
- Escort avoidance zone expands:
     - With each increment of time (nominally 10hz)
     - At a rate that is equal to or greater than the specified escort speed
     - In all drivable directions
- Escort avoidance zone has the potential to contract when:
     - AV receives escort position updates that allow the zone to be recalculated smaller

Reviewer 4: Updated in https://github.com/open-autonomy/escort/pull/5

- **Width**: Configured by the Escorter vehicle, applicable to escorting in open areas. When driving on road the escort must always stay within the lane, and the entire lane width will constitute the protection zone.

![Protection Zone Definition Image](images/escort-protection_zone.png)

### Real-Time Operating Conditions
Once an escort is activated, the Escorter vehicle continuously transmits its position at **1 Hz (one update per second)**. These periodic updates allow AVs to maintain an accurate escort zone. If updates are delayed or missed, the AV will automatically increase its Avoidance Zone to maintain safety margins to the Protection Zone. 
This adaptive expansion ensures AVs keep a protective buffer even when communication is unreliable.  

  .

> [!IMPORTANT]
> - The Escortee must at all times stay within the defined Protection Zone.
> - The AV will in its escort predictions assume that an Escort Convoy that is driving on a road will keep driving on the road and within the lane. Therefore it is important that the Escort Convoy actually does that.
---

## Escort Operation State Machine  
Escorts follow a lifecycle managed through the FMS–AHS communication:  

> - `Pending`: The escort operation has been announced but is not yet activated.  

Reviewer: "The escort operation has been announced but has not been activated by all operating AVs"

Reviewer 2: Suggestion https://github.com/open-autonomy/escort/pull/5

> - `Active`: The escort operation is active, transmitting location updates, and enforced by AVs.  

Reviewer: "The escort operation is active and enforced by all operating AVs."

The transmission of location updates is desired but not strictly required.

Reviewer 2: Suggestion https://github.com/open-autonomy/escort/pull/5

> - `Deleted`: The escort operation has been removed and is no longer enforced.  

Reviewer: "The escort operation has been removed and is no longer enforced by all operating AVs."

> 

Reviewer: - `Pending Delete`: The escort operation has been marked for deletion but has not yet been removed by all operating AVs.

```mermaid
stateDiagram-v2
state "Pending" as p
state "Pending Delete" as pd
state "Active" as a
state "Deleted" as d

[*] --> p : Create Escort in BM
p --> p : AVs report pending
p --> a : Escort Activated by all AVs
a --> pd : Escort Deactivated by Escorter
p --> pd : Escort Deactivated by Escorter

> p --> pd : An AV Rejected the Escort

Reviewer: AV rejects -> Deletes escort. TBD as this differs from the PZS structure.

PZS structure a 'Rejection' from an AV does not auto delete the zone. The zone remains in pending and the FMS operator takes action to decide how to proceed (delete the zone, retain partial activation & attempt to resolve the rejections from the AV/s).

The distinction between AV 'Pending' and AV 'Reject' would be:
- AVs that report pending are expected to self-resolve the issue preventing activation
- AVs that report rejected are expected to require some intervention before it is able to activate the escort

The question is, should the escort remain in pending while the issue is resolved, or should it just be deleted:
- Remaining in pending could cause production issues with no real benefit (because the escort isn't safe to use)
- Auto deleting could result in frustrating user experience

Either way, I believe this needs to be managed by the FMS, because it will need to synchronize this update with the rest of the fleet - i.e.:
 - FMS sends 'DeactivateEscortRequest' to all other AVs.

Due to this I think we shouldn't bake this into the state diagram.

pd --> d : Deactivated by all AVs

> d --> [*] : Escort Retired in BM

Reviewer: Remove reference to BM. Can replace with FMS.

Reviewer 2: Suggestion https://github.com/open-autonomy/escort/pull/5

```

> [!NOTE]  

> > Immutable attributes (e.g., escort vehicle ID, length and width [TBD - Station ID (V2X ID)], communication configuration) cannot be changed on an active zone. To update them, a new escort zone must be created, and the old one must be retired.  

Reviewer: For simplicity should we just say that escorts are immutable (at the object level rather than the param level)? All params synced over this interface should be immutable.

Any mutable params would need to be isolated to one side (i.e. the AHS or the FMS) so can be dealt with outside of this interface specification.

---

## Sequence Diagrams  
See [Sequence Diagrams](diagram/SequenceDiagrams.md) for scenarios illustrating the interactions between FMS, AHS, and AV when creating, activating, updating, and removing escort zones.  

---

## Communication Protocol  

> All communication between the FMS and AHS for managing escort zones is handled through a **single persistent WebSocket connection**.  

Reviewer: Should we be specifying the use of websockets in this standard, or should we attempt to keep the comms layer agnostic?

Reviewer 5: No, I believe you are correct. We should avoid specifying WebSockets in the standard. Keeping the communications layer agnostic allows for greater flexibility and future-proofing.

Reviewer 4: Updated. See https://github.com/open-autonomy/escort/pull/5

This connection must satisfy the following requirements:  
- The connection is continuously monitored from both ends.  
- Messages are exchanged asynchronously over the connection.  
- Escort position updates are streamed at **1 Hz** while the escort is pending or active.  
- If the connection is lost, the following applies:  
  - The connection must be reestablished **as soon as possible**.  

>   - During the outage, AVs automatically expand the Avoidance Zone to maintain safety.  

Reviewer: Shouldn't the Avoidance Zone always be expanding? Position updates make corrections if/when they are received, but they are not strictly necessary.

This will be necessary if we are using 2 potential sources for position updates - i.e. FMS, v2x.

Reviewer 2: See https://github.com/open-autonomy/escort/pull/5

  - Once the connection is restored, the Avoidance Zone contracts back to its configured dimensions.  

---

> ## Failure Handling  

Reviewer: The safety case for Escort is established at time of creation.

The position update mechanism/s reduces the impact to production and improves the usability of the system. Whether AVs receive position updates after the time of creation should have no bearing on the safety of the system.

Reviewer: I think we need a new section for Position Updates that describes this plus:
 - Update frequency
 - Multiple sources for position updates (optional)

Reviewer 2: Suggestion https://github.com/open-autonomy/escort/pull/5

### Deactivation
Escorter and Escortee(s) position at a safe staging location near or within the AOZ. Escorter issues an escort deactivation request. FMS distributes the escort deactivation to all active AVs.

## Protection Zone Specification
Protection Zones are parameterized during escort creation and are immutable for the active escort instance.

### Parameters
- Length: Maximum longitudinal trailing distance within which all Escortees must remain behind the Escorter along its driven path.
- Width: Lateral extent used in open areas. On roads, lane boundaries supersede Width and define the Protection Zone lateral limits.

![Protection Zone Definition](images/escort-protection_zone.png)

### Real‑Time Updates
Escorter position is transmitted nominally at 1 Hz (one update per second) while Pending or Active. AVs use these updates to constrain expansion of the Avoidance Zone.

## Prediction Model
Each AV shall predict possible Escorter positions between updates according to these assumptions:
- On roads:
  - Escorter speed may be any value up to OnRoadSpeedLimit defined in the activate request.
  - No reversing within a lane.
  - Lane adherence; no crossing into opposing lanes.

>   - At legal splits, any permitted branch may be chosen.

Reviewer 3: Is a "legal split" the same as an intersection? Should we re-phrase to make this obvious (e.g. "At legal splits (interstections), any ...

Reviewer 2: Yes it refers to intersections. I would say "legal" is meant to imply that it should be following allowed lane directions.

- In open areas:
  - Escorter speed may be any value up to the map‑defined maximum for the current open area.
  - No reversing.
  - Any forward trajectory consistent with vehicle dynamics within the configured Width and site constraints may be taken.

## Operation State Machine
Escort lifecycle states exchanged between FMS and AHS:
- Pending: Escort announced; activation acknowledgments outstanding.
- Active: Escort fully acknowledged; zone enforcement in effect.
- Deleted: Escort retired; zone no longer enforced.
- Pending Delete: The escort operation has been marked for deletion but has not yet been removed by all operating AVs.

### State Machine In FMS
```mermaid
stateDiagram-v2
state "Pending" as p
state "Pending Delete" as pd
state "Active" as a
state "Deleted" as d

[*] --> p : Create Escort in FMS
p --> p : AVs report pending
p --> a : Escort Activated by all AVs
a --> pd : Escort Deactivated by Escorter
p --> pd : Escort Deactivated by Escorter
p --> pd : An AV Rejected the Escort
pd --> d : Deactivated by all AVs

> d --> [*] : Escort Retired in FMS

Reviewer 3: ```suggestion
stateDiagram-v2
state "Pending" as p
state "Pending Delete" as pd
state "Active" as a
state "Deleted" as d

[*] --> p : Create Escort in FMS

p --> a : Escort Activated by all AVs
a --> pd : Escort Deactivated by Escorter
p --> pd : Escort Deactivated by Escorter
p --> p : An AV Rejects the Escort or AVs report pending
pd --> d : Deactivated by all AVs
d --> [*] : Escort Retired in FMS
```

p --> p : AV wants to postpone Activation
p --> a : AV activates escort
a --> a : Escort is active and AV avoids protection zone
p --> d : AV rejects escort
p --> d : Incoming DeactivateEscortRequest
a --> d : Incoming DeactivateEscortRequest
d --> [*] : Escort Removed from AV
```

> [!NOTE]
> Immutable attributes (Escorter vehicle ID, Length, Width, communication mode, and any future station identifiers) shall not be modified while Active. To change them, create a new escort and retire the existing one.

## Sequence Diagrams
Refer to [sequence diagrams](diagram/SequenceDiagrams.md) for interaction flows covering creation, activation, updates, and retirement.

## Communication Protocol
Escort management messages are exchanged between FMS and AHS.

Requirements:
- Continuous health monitoring by both endpoints.
- Asynchronous, non‑blocking message exchange.
- Position updates at 1 Hz during Pending and Active states.
- Rapid reconnection following any connection loss.

## Glossary Consistency Notes
- Use “Escorter” (capitalized) and “Escortee” consistently.
- “Protection Zone” refers to the convoy zone; “Avoidance Zone” is the AV internal safety buffer.

> - Normative verbs: “shall” for mandatory, “may” for optional capability, “should” for recommended optimization.

Reviewer 3: Do we need this? If so, should it just be part of the ##Terminology section?

Reviewer 2: We could probably remove this section.

---

> Copyright (c) Contributors. Licensed under the terms in `LICENSE`.

Reviewer 3: ```suggestion
Copyright (c) Contributors. Licensed under the terms in [LICENSE](LICENSE).
```

This repository defines the **messages and communication protocol between the Fleet Management System (FMS) and the Autonomous Haulage System (AHS)** for activating and deactivating escorts. Through this protocol, AVs in the area enforce a **no-drive zone** around the escort vehicle, ensuring safe movement for non-autonomous or non-equipped vehicles.  

---

### Language  
| Acronyms | Extended Name |  
| --- | --- |  
| AHS | Autonomous Haulage System |  
| AV | Autonomous Vehicle |  
| FMS | Fleet Management System |  

---

## Escort Zone Specification  
Escort zones are defined and exchanged using this protocol. The zones are dynamically shaped based on both **escort-defined parameters** and **real-time operating conditions**:  

- **Width**: Configured by the escorting vehicle, but constrained by the lane width.  
- **Length**: Initially defined by the escort, but may extend depending on communication latency and the escort’s speed.  

>   - If position updates are immediate, the zone length equals the escort vehicle length.  

Reviewer 6: ```suggestion
  - If position updates are immediate, the zone length equals the specified parameter Protection zone length.  
```
We need to look over the naming convention for Escort zone/no-drive zone/safety zone/protection area

  - If updates are delayed, the zone expands in proportion to the elapsed time and the escort’s speed.  

> 

Reviewer 2: Suggest to add an image similar to this:

<img width="1742" height="636" alt="escort-protection_zone" src="https://github.com/user-attachments/assets/333de5c2-e8ca-4368-bddc-b33590827096" />

**TODO**
Define reference point on escort lead vehicle. Any suggestions?

Reviewer 2: Or using nomenclature Escorter and Escortee:

<img width="1200" height="448" alt="escort-protection_zone" src="https://github.com/user-attachments/assets/b9a0b6aa-03f4-4b57-a335-76557a2ce726" />

> [!IMPORTANT]  
> Once an escort is activated, the escort vehicle continuously transmits its position at **1 Hz (one update per second)**. These periodic updates allow AVs to maintain an accurate escort zone. If updates are delayed or missed, the zone length automatically increases to maintain safety margins.  

This adaptive expansion ensures AVs keep a protective buffer even when communication is unreliable.  

---

## Escort Zone State Machine  
Escort zones follow a lifecycle managed through the FMS–AHS communication:  

- `Pending`: The escort zone has been created but not yet activated.  
- `Active`: The escort zone is active, transmitting location updates, and enforced by AVs.  
- `Deleted`: The escort zone has been removed and is no longer enforced.  

> ![Escort Zone State Machine](draw.io/EscortZoneStateMachine.svg)  

Reviewer 5: Broken link, maybe an svg needs to be created similar to zone repo.

Reviewer 2: Created a mermaid diagram here https://github.com/open-autonomy/escort/pull/3/commits/4383ac0111ff227d609c20c24eff78a498531a9b not as good looking though..

| Escort Protection Zone | The zone in which the escorting convoy is required to be within at all times. Is defined by a configurable distance that extends behind the escorter, to include the escortee. |
| Escort Avoidance Zone | Zone internally defined on each AV, that the AV will not enter. It is at least as big as the protection zone, but is likely bigger due to latency of the escorter's position updates. |

---

## Escort Protection Zone Specification  
Escort protection zones are defined and exchanged using this protocol. The zones are dynamically shaped based on both **escort-defined parameters** and **real-time operating conditions**:

### Parameters
These parameters are part of the activation request, during the active escort phase the parameters are constant througout the entire escort operation.
- **Length**: The Escortee must always stay within _length_ meters of the Escorter along the path that the Escorter is driving.
- **Width**: Configured by the Escorter vehicle, applicable to escorting in open areas. When driving on road the escort must always stay within the lane, and the entire lane width will constitute the protection zone.

> ![Protection Zone Definition Image](images/escort-protection_zone.png)

Reviewer 5: link seems to be broken

Reviewer 4: Seems to work for me and Gustav
